File: stage2/cpu/idt.c
#include "idt.h"

// Apparently aligning gives performance? https://wiki.osdev.org/Interrupts_Tutorial
__attribute__((aligned(0x10)))
static struct IDTEntry kIDTTable[256];

static struct IDTDescriptor kIDTDescriptor = {
    .idt = kIDTTable,
    .size = sizeof(struct IDTEntry) * 256 - 1,
};

void IDTInsertEntry(uint8_t index, struct IDTEntry entry) {
    kIDTTable[index] = entry;
}

struct IDTEntry IDTMakeEntry(void(*isr)(void), union IDTEntryFlags flags) {
    struct IDTEntry entry = {
        .isr_low = (uint32_t)isr & 0xFFFF,
        .kernel_cs = 0x08,
        .reserved = 0,
        .flags = flags,
        .isr_high = (uint32_t)isr >> 16,
    };

    return entry;
}

struct InterruptRegisters {
    uint32_t edi, esi, ebp, esp, ebx, edx, ecx, eax; // Order pusha pushes registers in
    uint32_t interrupt_number, error_code; // Pushed in isr.asm
    uint32_t eip, cs, eflags; // Pushed by CPU https://wiki.osdev.org/Interrupt_Service_Routines
};

char* exception_messages[32] = {
    "Division By Zero",
    "Debug",
    "Non Maskable Interrupt",
    "Breakpoint",
    "Into Detected Overflow",
    "Out of Bounds",
    "Invalid Opcode",
    "No Coprocessor",
    "Double Fault",
    "Coprocessor Segment Overrun",
    "Bad TSS",
    "Segment Not Present",
    "Stack Fault",
    "General Protection Fault",
    "Page Fault",
    "Unknown Interrupt",
    "Coprocessor Fault",
    "Alignment Check",
    "Machine Check",
    "Reserved",
    "Reserved",
    "Reserved",
    "Reserved",
    "Reserved",
    "Reserved",
    "Reserved",
    "Reserved",
    "Reserved",
    "Reserved",
    "Reserved",
    "Reserved",
    "Reserved"
};


void CISRHandler(struct InterruptRegisters registers) {
    TerminalWriteString("Interrupt Message: ");
    TerminalWriteString(exception_messages[registers.interrupt_number]);
    for (;;);
}

#define EXCEPTION(num) extern void exception##num(void)

EXCEPTION(0); EXCEPTION(1); EXCEPTION(2); EXCEPTION(3);
EXCEPTION(4); EXCEPTION(5); EXCEPTION(6); EXCEPTION(7);
EXCEPTION(8); EXCEPTION(9); EXCEPTION(10); EXCEPTION(11);
EXCEPTION(12); EXCEPTION(13); EXCEPTION(14); EXCEPTION(15);
EXCEPTION(16); EXCEPTION(17); EXCEPTION(18); EXCEPTION(19);
EXCEPTION(20); EXCEPTION(21); EXCEPTION(22); EXCEPTION(23);
EXCEPTION(24); EXCEPTION(25); EXCEPTION(26); EXCEPTION(27);
EXCEPTION(28); EXCEPTION(29); EXCEPTION(30); EXCEPTION(31);

void IDTInit(void) {
    struct IDTEntry entry = IDTMakeEntry(exception0, (union IDTEntryFlags){
        .reserved = 0,
        .cpu_priviledge_level = 0,
        .gate_type = kInterruptGate32,
        .present = 1,
    });

    for (uint8_t i = 0; i < 32; i++) {
        IDTInsertEntry(i, entry);
    }
    
    __asm__ volatile ("lidt %0" :: "m"(kIDTDescriptor));
    __asm__ volatile ("sti");
}

File: stage2/arch/bios/asm/idt.zig
const std = @import("std");

const isr = @import("isr.zig");
const frame = @import("frame.zig");
const debug = @import("../debug.zig");

pub const GateType = enum(u4) {
    task_gate = 0x5,
    interrupt_gate_16 = 0x6,
    trap_gate_16 = 0x7,
    interrupt_gate_32 = 0xE,
    trap_gate_32 = 0xF,
};

// Format can be found at https://wiki.osdev.org/IDT
// Flags: PDD0GGGG
// - P is the Present Bit, must be set to 1 for all valid entries
// - D is the Cpu Priviledge Level needed to use INT to use this, 0-3
// - G is the Gate Type
pub const Flags = packed union {
    flags: u8,
    specific: packed struct {
        gate_type: GateType = .trap_gate_32,
        reserved: u1 = 0,
        priviledge_level: u2 = 0,
        present: bool = true,
    },

    comptime {
        std.debug.assert(@sizeOf(@This()) == 1);
    }
};

pub const Entry = packed struct {
    isr_low: u16,
    kernel_cs: u16 = 0x8,
    reserved: u8 = 0,
    flags: Flags = .{ .specific = .{} },
    isr_high: u16,
};

pub const Descriptor = packed struct {
    size: u16,
    pointer: usize,
};

pub const ISRHandler = *const fn (isr.InterruptInfo) callconv(.C) void;

// zig fmt: off
fn getExceptionMessage(vec: usize) []const u8 {
    return switch (vec) {
        0 => "Division By Zero",
        1 => "Debug",
        2 => "Non Maskable Interrupt", 
        3 => "Breakpoint", 
        4 => "Into Detected Overflow", 
        5 => "Out of Bounds", 
        6 => "Invalid Opcode", 
        7 => "No Coprocessor", 
        8 => "Double Fault", 
        9 => "Coprocessor Segment Overrun", 
        10 => "Bad TSS", 
        11 => "Segment Not Present", 
        12 => "Stack Fault", 
        13 => "General Protection Fault", 
        14 => "Page Fault", 
        15 => "Unknown Interrupt", 
        16 => "Coprocessor Fault", 
        17 => "Alignment Check", 
        18 => "Machine Check", 
        19...31 => "Reserved",
        else => "Unknown"
    };
}
// zig fmt: on

// Aligning gives performance https://wiki.osdev.org/Interrupts_Tutorial
var table: [256]Entry align(0x10) = undefined;

// Stores the functions that are run on an interrupt
pub var func_table: [256]ISRHandler = undefined;

pub fn makeEntry(func: isr.Stub) Entry {
    const int_addr = @intFromPtr(func);

    const entry: Entry = .{
        .isr_low = @truncate(int_addr),
        .isr_high = @truncate(int_addr >> 16),
    };

    return entry;
}

pub fn defaultISRHandler(regs: isr.InterruptInfo) callconv(.C) void {
    std.log.info("ISR Info Dump:", .{});
    regs.dump();
    std.log.info("Frame Dump:", .{});
    const f = frame.getFrame();
    f.dump();
    @panic(getExceptionMessage(regs.interrupt_number));
}

pub fn installInterrupt(comptime vec: u8, func: ISRHandler) void {
    func_table[vec] = func;
    table[vec] = makeEntry(isr.makeStub(vec));
}

pub fn init() void {
    inline for (0..32) |vec| {
        installInterrupt(vec, defaultISRHandler);
    }

    const descriptor = Descriptor{
        .pointer = @intFromPtr(&table),
        .size = @sizeOf(Entry) * 256,
    };

    const idtr = @intFromPtr(&descriptor);

    asm volatile ("lidt (%[idtr])"
        :
        : [idtr] "r" (idtr),
    );
}
